src/main/java/dev/galacticraft/mod/api/block/PipeShapedBlock.java

```diff
@@ -0,0 +1,180 @@
+    public final VoxelShape[] shapes;
+
+    protected PipeShapedBlock(float radius, BlockBehaviour.Properties properties) {
+        super(properties);
+        this.shapes = makeShapes(radius);
+    }
+
+    @Override
+    abstract public @Nullable BE newBlockEntity(BlockPos pos, BlockState state);
+
+    abstract public boolean canConnectTo(Level level, BlockPos thisPos, Direction direction, BlockPos neighborPos, BlockState thisState);
+
+    abstract protected void onConnectionChanged(Level level, BlockPos thisPos, Direction direction, BlockPos neighborPos);
+
+    protected boolean updateConnection(BlockState currentState, BlockPos pos, Direction side, BlockPos neighborPos, Level level) {
+        if (level.getBlockEntity(pos) instanceof Connected pipe) {
+            boolean canConnect = this.canConnectTo(level, pos, side, neighborPos, currentState);
+
+            BlockState neighborState = level.getBlockState(neighborPos);
+            if (neighborState.getBlock() instanceof PipeShapedBlock<?> neighbor) {
+                canConnect &= neighbor.canConnectTo(level, neighborPos, side.getOpposite(), pos, neighborState);
+            }
+
+            boolean currentlyConnected = pipe.getConnections()[side.get3DDataValue()];
+            pipe.getConnections()[side.get3DDataValue()] = canConnect;
+            level.blockEntityChanged(pos);
+            return canConnect != currentlyConnected;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void setPlacedBy(Level level, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack itemStack) {
+        super.setPlacedBy(level, pos, state, placer, itemStack);
+
+        for (Direction direction : Direction.values()) {
+            this.updateConnection(state, pos, direction, pos.relative(direction), level);
+        }
+    }
+
+    @Override
+    public void neighborChanged(BlockState state, Level level, BlockPos pos, Block block, BlockPos neighborPos, boolean notify) {
+        super.neighborChanged(state, level, pos, block, neighborPos, notify);
+
+        Direction direction = Direction.fromDelta(neighborPos.getX() - pos.getX(), neighborPos.getY() - pos.getY(), neighborPos.getZ() - pos.getZ());
+        if (direction == null)
+            return;
+
+        if (this.updateConnection(state, pos, direction, neighborPos, level)) {
+            level.updateNeighborsAtExceptFromFacing(pos, state.getBlock(), direction);
+            this.onConnectionChanged(level, pos, direction, neighborPos);
+        }
+    }
+
+    @Override
+    protected @NotNull BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor levelAccessor, BlockPos pos, BlockPos neighborPos) {
+        if (levelAccessor instanceof Level level) {
+            if (this.updateConnection(state, pos, direction, neighborPos, level)) {
+                level.updateNeighborsAtExceptFromFacing(pos, state.getBlock(), direction);
```

Comment:
Same thing again, but with the added caveat that this additional update/check might be redundant if the `neighborChanged` check updates the changed direction.